app/models/manageiq/providers/azure/cloud_manager/metrics_capture.rb

```diff
@@ -77,76 +77,135 @@ def perf_collect_metrics(interval_name, start_time = nil, end_time = nil)
-          timestamp = Time.zone.parse(metric_value.time_stamp)
-          (ts_memo[timestamp] ||= []) << metric_value.average
+      storage_conn = Azure::Armrest::StorageAccountService.new(@connection)
+      storage_accounts = storage_conn.list_all
+
+      metric_definitions, _timings = Benchmark.realtime_block(:capture_counters) do
+        # azure-armrest gem needs to be modified to accept version ('list_definitions' method)
+        response = metrics_conn.send(:rest_get, definitions_url.to_s)
+        Azure::Armrest::ArmrestCollection.create_from_response(response, Azure::Armrest::Insights::MetricDefinition)
+      end
+
+      metric_definitions.each do |metric_definition|
+        metric_name = metric_definition.name.value
+        next unless METRIC_NAMES[:raw].include?(metric_name)
+
+        metric_availability = metric_definition.metric_availabilities.detect { |ma| ma.time_grain == INTERVAL_1_MINUTE }
+        next unless metric_availability
+
+        timestamped_values = metrics[metric_name] ||= {}
+
+        metric_location = metric_availability.location
+
+        storage_account_name = URI.parse(metric_location.table_endpoint).host.split('.').first
+        storage_account      = storage_accounts.find { |account| account.name == storage_account_name }
+        storage_account_keys = storage_conn.list_account_keys(storage_account.name, storage_account.resource_group)
+        storage_account_key  = storage_account_keys.fetch('key1')
+
+        metric_location.table_info.each do |table_info|
+          t_start_time = parse_timestamp(table_info.start_time)
+          t_end_time   = parse_timestamp(table_info.end_time).end_of_day
+          next unless time_interval.overlaps?(t_start_time..t_end_time)
+
+          filter = [
+            "CounterName eq '#{metric_name}'",
+            "PartitionKey eq '#{metric_location.partition_key}'",
+            "Timestamp ge datetime'#{@start_time}'",
+            "Timestamp le datetime'#{@end_time}'",
+          ].join(' and ')
+
+          table_data, _timings = Benchmark.realtime_block(:capture_counters) do
+            storage_account.table_data(
+              table_info.table_name,
+              storage_account_key,
+              :filter => filter,
+              :select => 'TIMESTAMP,Average',
+              :all    => true
+            )
+          end
+
+          table_data.each do |row_data|
+            timestamp = parse_timestamp(row_data.timestamp)
+            (timestamped_values[timestamp] ||= []) << row_data.average
+          end
         end
+      end
 
-        next if timestamped_values.empty?
+      ## Organize data in the proper form
+
+      COUNTERS_INFO.values.flatten.each do |counter_info|
+        metric_values = metrics.values_at(*counter_info.metrics).compact
+        next if metric_values.empty?
+
+        # { timestamp => [value, ...], ... }
+        timestamped_values = metric_values.each_with_object({}) do |ts_values, memo|
```

Comment:
Just a style thing but could this be simplified to something like

```
timestamped_values = Hash.new { |h, k| h[k] = [] }
metric_values.each { |ts, values| timestamped_values[ts].concat(values) }
```

Reply:
If metric_values is already a hash with the timestamp as the key then what exactly is this whole thing transforming it from and to?

Reply:
it transforms this:
```
{
  metric_name_1 => {
    timestamp1 => [value1, value2, ...],
    timestamp2 => ...
  },
  metric_name_2 => {
    timestamp1 => [value3, value4, ...],
    timestamp3 => ...
  },
  ...
}
```
into this:
```
{
  timestamp1 => [value1, value2, value3, value4],
  timestamp2 => ...
  ...
}
```